docs/madr/decisions/020-kds-redesign.md

```diff
@@ -0,0 +1,247 @@
+
+Maybe not in the first step but we could consider if we can benefit from using different cache implementation `LinearCache` or `MuxCache` for snapshot generation.
+
+### Prioritize operational informations
+
+Certain resources require more frequent synchronization than others. For example, Policies, ZoneIngress, and ZoneEgress must be updated immediately, whereas Insights and Dataplanes objects may be delayed and can be aggregated. To address this, we could consider implementing an option to specify the interval at which the snapshot for specific resources is recalculated. To avoid recalculating changes for all synchronized resources, we need a cache per resource type, which `SimpleCache` should allows us to achieve. In addition, we could set up an RPC service to monitor changes for all resources on separate streams, allowing changes to be processed independently.
+
+### Change the namespace of synchronized resources from `default` to `kuma-system`
+
+Resources synchronized from Zone to Global are created by default in a `default` namespace and we should change it in the new version. This is a breaking change because upgrade won't be able to find resources and needs to duplicate them from zone and store them in the `kuma-system` namespace.
+
+### Handshake
+
+Is it possible for the zone control-plane to be newer than the global control-plane?
+
+If the answer is no, then a handshake is not necessary as the client will use the latest available version.
+
+If the answer is yes, then a handshake API may need to be implemented. The GRPC Reflection API allows services to be discovered on the server side, but not on the client side.
+
+The output of the reflection functionality is:
+
+```
+["grpc.reflection.v1alpha.ServerReflection", "kuma.mesh.v1alpha1.GlobalKDSService", "kuma.mesh.v1alpha1.MultiplexService"]```
+```
+
+How would a handshake work?
+
+1. The server exposes the RPC `handshake`.
+2. The client initiates a connection to the server.
+3. The client sends a request that includes the Kuma version and the supported KDS versions:
+```
+ kumaVersion: "2.1.0",
+ supportedKdsVersions: ["0.1.0", "0.2.0"],
+```
+4. Server checks the request and responds with the newest version working for both
+```
+ kdsVersion: "0.2.0"
+```
+5. The client needs to create RPC clients for the specific version.
+
+### Multitenant
+
+Global control-plane should support multitenant, that means we want to achieve that one global control-plane can handle many different zone control-planes that are different tenant. One approach to accomplish this is by extracting tenant specifics from an authentication token. These details can then be transmitted along with a request to indicate the specific tenant and retrieve solely their deployment information. Another option is to supply this information via a header. It appears that accommodating multitenancy doesn't require additional work on the KDS side.
```

Comment:
Sorry, it's probably me not understanding things, but why it says "zone control-planes that are different tenant"? I thought we agreed that Global CP is multitenant (one instance of Global CP is acting as many Global CP instances), but Zone CP is always single-tenant (like today), right?